Add {floor,ceil}_char_boundary methods to str

clarfonthey opened the pull request:
This is technically already used internally by the standard library in the form of `truncate_to_char_boundary`.

Essentially these are two building blocks to allow for approximate string truncation, where you want to cut off the string at "approximately" a given length in bytes but don't know exactly where the character boundaries lie. It's also a good candidate for the standard library as it can easily be done naively, but would be difficult to properly optimise. Although the existing code that's done in error messages is done naively, this code will explicitly only check a window of 4 bytes since we know that a boundary must lie in that range, and because it will make it possible to vectorise.

Although this method doesn't take into account graphemes or other properties, this would still be a required building block for splitting that takes those into account. For example, if you wanted to split at a grapheme boundary, you could take your approximate splitting point and then determine the graphemes immediately following and preceeding the split. If you then notice that these two graphemes could be merged, you can decide to either include the whole grapheme or exclude it depending on whether you decide splitting should shrink or expand the string.

This takes the most conservative approach and just offers the raw indices to the user, and they can decide how to use them. That way, the methods are as useful as possible despite having as few methods as possible.

(Note: I'll add some tests and a tracking issue if it's decided that this is worth including.)

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2870345234) [(plain)](https://github.com/rust-lang/rust/commit/379b542732118fa41da308122ca8de54a9fd804e/checks/2870345234/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libc v0.2.93
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.45
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0599]: no method named `rfind` found for slice `[u8]` in the current scope
    |
    |
245 |             let index = self.as_bytes()[lower_bound..=index].rfind(
    |                                                              ^^^^^ method not found in `[u8]`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `rfind`, perhaps you need to implement one of them:
            candidate #1: `FuseImpl`
            candidate #2: `DoubleEndedIterator`

error[E0599]: no method named `rfind` found for slice `[u8]` in the current scope
    |
    |
283 |             let index = self.as_bytes()[index..upper_bound].rfind(
    |                                                             ^^^^^ method not found in `[u8]`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `rfind`, perhaps you need to implement one of them:
            candidate #1: `FuseImpl`
            candidate #2: `DoubleEndedIterator`
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0599`.
error: could not compile `core`
error: could not compile `core`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:04
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2870359544) [(plain)](https://github.com/rust-lang/rust/commit/8699967e68166b3766ccdcd9c5f2ddab76fa754e/checks/2870359544/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0308]: mismatched types
   --> library/core/src/str/mod.rs:249:22
    |
241 |     pub fn nearest_char_boundary_max(&self, index: usize) -> usize {
    |                                                              ----- expected `usize` because of return type
...
249 |             unsafe { index.unwrap_unchecked() }
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^ expected `usize`, found `&u8`

error[E0606]: casting `&&u8` as `i8` is invalid
    |
    |
246 |             let index = self.as_bytes()[lower_bound..=index].iter().rfind(|b| (b as i8) >= -0x40);
    |
    |
    = help: cast through a raw pointer first
error[E0308]: mismatched types
   --> library/core/src/str/mod.rs:284:29
    |
    |
284 |             index.unwrap_or(upper_bound)
    |                             ^^^^^^^^^^^ expected `&u8`, found `usize`
error[E0308]: mismatched types
   --> library/core/src/str/mod.rs:284:13
    |
    |
276 |     pub fn nearest_char_boundary_min(&self, index: usize) -> usize {
    |                                                              ----- expected `usize` because of return type
...
284 |             index.unwrap_or(upper_bound)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `usize`, found `&u8`

error[E0606]: casting `&&u8` as `i8` is invalid
    |
    |
281 |             let index = self.as_bytes()[index..upper_bound].iter().find(|b| (b as i8) >= -0x40);
    |
    |
    = help: cast through a raw pointer first
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0308, E0606.
For more information about an error, try `rustc --explain E0308`.
```

</details>

clarfonthey commented:
(I know how to fix the error but my desktop is in literal pieces at the moment, so, I'll fix it this weekend.)

scottmcm commented:
The functionality of these seems like a good thing to provide :+1:

For the bikeshed, mixing "nearest" and "min/max" in the name is a bit confusing to me.  Maybe `char_boundary_begin` and `char_boundary_end` or something?  (And thus `is_char_boundary == (char_boundary_begin == char_boundary_end)`.)

Or maybe, with subtly different semantics `char_{begin|end}`?  So that `s.char_begin(i) .. s.char_end(i)` is the `Range<usize>` representing the USV that includes `s.as_bytes()[i]`?

That raises the question of endpoints, I guess.  What's `.nearest_char_boundary_min(usize::MAX)` supposed to do?

clarfonthey commented:
> That raises the question of endpoints, I guess.  What's `.nearest_char_boundary_min(usize::MAX)` supposed to do?

Forgetting the names (and I agree, it's hard to come up with something reasonable) the current behaviour here is to panic.

The one which is upper-bounded will currently not panic, which I think is good for truncating behaviour -- if you want your string to be under, say, 20 bytes, and it's empty, that's fine.

Lower-bounded ones must panic if the index is above len, since they can't go any lower.

If we were to use the concept of bounding the results to a specific character, then our issue is that we treat past the end as an "end of string" character, and this isn't C.

My initial terms for the methods were "largest char boundary not exceeding" and "smallest char boundary not below," and I figured those names sucked, but that's the intent that I wanted to preserve in the methods themselves.

Mark-Simulacrum commented on library/core/src/str/mod.rs:
```diff
@@ -217,6 +216,75 @@ impl str {
+            let lower_bound = index.saturating_sub(3);
+            let index = self.as_bytes()[lower_bound..=index].iter().rfind(|b| (b as i8) >= -0x40);
+
+            // SAFETY: we know that the character boundary will be within four characters
```
```suggestion
            // SAFETY: we know that the character boundary will be within four bytes
```

Mark-Simulacrum commented on library/core/src/str/mod.rs:
```diff
@@ -217,6 +216,75 @@ impl str {
+        if index > self.len() {
+            slice_error_fail(self, index, index)
+        } else {
+            let upper_bound = (index + 4).max(self.len());
```
```suggestion
            let upper_bound = (index + 4).min(self.len());
```

Since we want to not exceed the length, right? `.clamp(0, self.len())` might be better, not sure if it optimizes out clamping above zero though.

clarfonthey replied:
Hmm, you are right about clamping; I'll test that the >= 0 is elided but it should be.

clarfonthey replied:
It is: https://rust.godbolt.org/z/hcTsahs14

Mark-Simulacrum commented:
r? @dtolnay as T-libs-api to evaluate the addition of unstable API surface area. Implementations look OK to me but I didn't evaluate the correctness of the search criteria (i.e., as casts and equality to -0x40).

clarfonthey commented:
Note that the -0x40 was taken directly from `is_char_boundary`. I could factor out that bit into a separate private method (like `byte_is_char_boundary`) to avoid potential drift.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2937817189) [(plain)](https://github.com/rust-lang/rust/commit/d2b0e71b68fc40138a0783f281354d1be950870f/checks/2937817189/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 625 error codes
* highest error code: E0783
tidy error: /checkout/library/core/src/num/mod.rs:687: tab character
Found 0 error codes with no tests
Done!
* 339 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:14
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2938226236) [(plain)](https://github.com/rust-lang/rust/commit/9616038199d3f76a7abad7ba21698c1b5f1f4010/checks/2938226236/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/core/src/str/mod.rs at line 243:
             self.len()
         } else {
             let lower_bound = index.saturating_sub(3);
-            let index = self.as_bytes()[lower_bound..=index].iter().rposition(|b| b.is_utf8_char_boundary());
+            let index = self.as_bytes()[lower_bound..=index]
+                .iter()
+                .rposition(|b| b.is_utf8_char_boundary());
 
             // SAFETY: we know that the character boundary will be within four bytes
             unsafe { index.unwrap_unchecked() }
Diff in /checkout/library/core/src/str/mod.rs at line 278:
             slice_error_fail(self, index, index)
         } else {
             let upper_bound = (index + 4).clamp(0, self.len());
-            let index = self.as_bytes()[index..upper_bound].iter().position(|b| b.is_utf8_char_boundary());
+            let index =
+                self.as_bytes()[index..upper_bound].iter().position(|b| b.is_utf8_char_boundary());
 
             // this case should only be reached if `upper_bound == self.len()`
             index.unwrap_or(upper_bound)
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/array/iter.rs" "/checkout/library/core/src/slice/raw.rs" "/checkout/library/core/src/char/convert.rs" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/str/lossy.rs" "/checkout/library/core/src/char/methods.rs" "/checkout/library/core/src/str/mod.rs" "/checkout/library/core/src/slice/index.rs"` failed.
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2938338833) [(plain)](https://github.com/rust-lang/rust/commit/0b94cfd011c6293cc9cc3df51e6c0f8d971b3206/checks/2938338833/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
 Documenting core v0.0.0 (/checkout/library/core)
error: unresolved link to `neraest_char_boundary_max`
    |
    |
257 |     /// This method is the natural complement to [`neraest_char_boundary_max`]. See that method
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `neraest_char_boundary_max` in scope
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
error: aborting due to previous error

error: could not document `core`


Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.55.0 --index-page /checkout/src/doc/index.md -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --cfg=bootstrap -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.55.0-nightly
  (961fa7011
  2021-06-29)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "core" "-Zskip-rustdoc-fingerprint" "--" "--markdown-css" "rust.css" "--markdown-no-toc" "-Z" "unstable-options" "--resource-suffix" "1.55.0" "--index-page" "/checkout/src/doc/index.md"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap doc --stage 0 library/test
Build completed unsuccessfully in 0:00:06
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2938414716) [(plain)](https://github.com/rust-lang/rust/commit/77823d8fb47f424ccd399e04cd2e4cd30bbc917e/checks/2938414716/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 600/658
..........................................................
failures:

---- str::slice_index::test_slice_fail_truncated_2 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_2' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:723:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"luctus, im`[...]"`
---- str::slice_index::test_slice_fail_truncated_1 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_1' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:717:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"byte index 1024 is out of bounds of `Lorem ipsum dolor sit amet"`
failures:
    str::slice_index::test_slice_fail_truncated_1
    str::slice_index::test_slice_fail_truncated_2
error: test failed, to rerun pass '-p alloc --test collectionstests'
error: test failed, to rerun pass '-p alloc --test collectionstests'

test result: FAILED. 656 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.50s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "alloc" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:17:09
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2944364322) [(plain)](https://github.com/rust-lang/rust/commit/c6f504bf1c0a9521b71b0e762c6560b70b113307/checks/2944364322/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 600/658
..........................................................
failures:

---- str::slice_index::test_slice_fail_truncated_1 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_1' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:717:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"byte index 1024 is out of bounds of `Lorem ipsum dolor sit amet"`
---- str::slice_index::test_slice_fail_truncated_2 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_2' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:723:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"luctus, im`[...]"`
error: test failed, to rerun pass '-p alloc --test collectionstests'
failures:
    str::slice_index::test_slice_fail_truncated_1
    str::slice_index::test_slice_fail_truncated_2


test result: FAILED. 656 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.49s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "alloc" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:17:04
```

</details>

clarfonthey commented:
(That test passed for me locally, and I'm not 100% sure on how to fix it. I also know that u8::is_utf8_char_boundary should probably be inlined. Mostly want to wait for a more affirmative "yes, I like this API ± name bikeshedding" before I put more work into it.

bluss commented on library/core/src/str/mod.rs:
```diff
@@ -75,15 +75,14 @@ use iter::MatchIndicesInternal;
+    let ellipsis = if trunc_len < s.len() { "[...]" } else { "" };
 
     // 1. out of bounds
     if begin > s.len() || end > s.len() {
```
On line 107 in this file there's another char boundary search snippet that could be replaced.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2967155461) [(plain)](https://github.com/rust-lang/rust/commit/8b6d128d84b5c055f0d3ec9625810db5747078e2/checks/2967155461/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

running 658 tests
.................................................................................................... 100/658
.................................................................................................... 200/658
...........................................F..FFFF.FF.FFFF.F.F.....F................................ 300/658
.F.............F.................................................................................... 400/658
.................................................................................................... 600/658
..........................................................
failures:


---- str::slice_index::boundary::range_1::index_mut_fail stdout ----
thread 'str::slice_index::boundary::range_1::index_mut_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::range_2::index_fail stdout ----
thread 'str::slice_index::boundary::range_2::index_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangefrom::index_fail stdout ----
thread 'str::slice_index::boundary::rangefrom::index_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
error: test failed, to rerun pass '-p alloc --test collectionstests'
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::range_1::index_fail stdout ----
thread 'str::slice_index::boundary::range_1::index_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::range_2::index_mut_fail stdout ----
thread 'str::slice_index::boundary::range_2::index_mut_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangefrom::index_mut_fail stdout ----
thread 'str::slice_index::boundary::rangefrom::index_mut_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::rangeinclusive_1::index_mut_fail stdout ----
thread 'str::slice_index::boundary::rangeinclusive_1::index_mut_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::rangeinclusive_1::index_fail stdout ----
thread 'str::slice_index::boundary::rangeinclusive_1::index_fail' panicked at 'byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 4 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 4 is not a char boundary; it is inside 'α' (bytes 3..5) of"`
---- str::slice_index::boundary::rangeinclusive_2::index_fail stdout ----
thread 'str::slice_index::boundary::rangeinclusive_2::index_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangeinclusive_2::index_mut_fail stdout ----
thread 'str::slice_index::boundary::rangeinclusive_2::index_mut_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangeto::index_fail stdout ----
thread 'str::slice_index::boundary::rangeto::index_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangeto::index_mut_fail stdout ----
thread 'str::slice_index::boundary::rangeto::index_mut_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangetoinclusive::index_fail stdout ----
thread 'str::slice_index::boundary::rangetoinclusive::index_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::boundary::rangetoinclusive::index_mut_fail stdout ----
thread 'str::slice_index::boundary::rangetoinclusive::index_mut_fail' panicked at 'byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`', library/alloc/tests/str.rs:653:9
note: panic did not contain expected string
      panic message: `"byte index 6 is not a char boundary; it is inside 'c' (bytes 2..3) of `abcαβγ`"`,
 expected substring: `"byte index 6 is not a char boundary; it is inside 'β' (bytes 5..7) of"`
---- str::slice_index::test_slice_fail_truncated_2 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_2' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:723:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"luctus, im`[...]"`
---- str::slice_index::test_slice_fail_truncated_1 stdout ----
thread 'str::slice_index::test_slice_fail_truncated_1' panicked at 'byte index 1024 is out of bounds of `Lor`[...]', library/alloc/tests/str.rs:717:18
note: panic did not contain expected string
      panic message: `"byte index 1024 is out of bounds of `Lor`[...]"`,
 expected substring: `"byte index 1024 is out of bounds of `Lorem ipsum dolor sit amet"`
failures:
    str::slice_index::boundary::range_1::index_fail
    str::slice_index::boundary::range_1::index_mut_fail
    str::slice_index::boundary::range_2::index_fail
---
test result: FAILED. 642 passed; 16 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.51s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "alloc" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:14:03
```

</details>

clarfonthey commented:
So, it turns out that for some reason the doc tests weren't running when I ran earlier. I found the error and I'm facepalming.

Anyway, tests should be fixed now.

Dylan-DPC commented:
@dtolnay this is ready for review

camsteffen commented:
Name ideas:
* `first_char_boundary_after`/`last_char_boundary_before`
* `find_char_boundary`/`rfind_char_boundary`
* `char_boundary_before`/`char_boundary_after` (I think I like this one)

clarfonthey commented:
Reason why I avoided just saying "before" and "after" is because they imply exclusivity, whereas the chosen index is also valid

clarfonthey commented:
ping @dtolnay, this PR has been hanging for a while and I'm not sure if it slid out of your radar, or if it might be better for someone else to review it. :)

bluss commented on library/core/src/str/mod.rs:
````diff
@@ -212,7 +208,81 @@ impl str {
+    /// ```
+    #[unstable(feature = "nearest_char_boundary", issue = "none")]
+    #[inline]
+    pub fn nearest_char_boundary_min(&self, index: usize) -> usize {
````
I think there are slightly less awkward names, reading the min/max names is not very intuitive unfortunately. I'm thinking of something like "next char boundary" or similar. I'm also sorry for bringing bikeshedding into this.

Maybe something like this pair:

- `find_char_boundary` ?
- `rfind_char_boundary` ?

Possibly the methods should have another word in the end, like `at` or `from`? Or we use `next`/`prev` as "verbs", but the non-inclusive connotation of those is a problem, similar to what you mentioned.

clarfonthey replied:
The bikeshedding is welcome; as I've said I haven't the slightest idea how to best name these.

`find` and `rfind` seem closer but I kind of want to emphasise that they're used for snapping to boundaries, rather than finding them.

clarfonthey replied:
Maybe char_boundary_floor and char_boundary_ceil?

bluss replied:
Not bad :)

clarfonthey commented:
Based upon the idea I had, I'm naming these `floor_char_boundary` and `ceil_char_boundary` because they can be thought of like a floor and ceil operation. Still sounds awkward but seems like the best names so far.

Would prefer to avoid bikeshedding before merge, though.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3641422997) [(plain)](https://github.com/rust-lang/rust/commit/a9da594b6a18d4565c7c4d38a256fd256c9f1554/checks/3641422997/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.17s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
 Documenting core v0.0.0 (/checkout/library/core)
error: unresolved link to `floor_char_boundary`
    |
    |
253 |     /// This method is the natural complement to [`floor_char_boundary`]. See that method
    |                                                    ^^^^^^^^^^^^^^^^^^^ no item named `floor_char_boundary` in scope
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
error: could not document `core`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.57.0 --index-page /checkout/src/doc/index.md -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --cfg=bootstrap -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.57.0-nightly
  (d75c1ace5
  2021-09-18)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "core" "-Zskip-rustdoc-fingerprint" "--" "--markdown-css" "rust.css" "--markdown-no-toc" "-Z" "unstable-options" "--resource-suffix" "1.57.0" "--index-page" "/checkout/src/doc/index.md"


Build completed unsuccessfully in 0:00:06
```

</details>

camsteffen commented:
That is actually quite intuitive IMO. 👍

camsteffen commented on library/core/src/str/mod.rs:
```diff
@@ -212,7 +208,81 @@ impl str {
+        }
+    }
+
+    /// Finds the closest `x` not exceeding `index` where `is_char_boundary(x)` is `true`.
```
Just a suggestion...

```suggestion
    /// Finds the nearest char boundary at or below `index`.
    ///
    /// See the [is_char_boundary] method.
```

bors commented:
:umbrella: The latest upstream changes (presumably #88834) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

clarfonthey commented:
Will try to rebase soon.

falk-hueffner commented on library/core/src/str/mod.rs:
```diff
@@ -213,7 +209,81 @@ impl str {
             None => index == self.len(),
 
             // This is bit magic equivalent to: b < 128 || b >= 192
```
This comment should probably move to `is_utf8_char_boundary`.

dtolnay commented on library/core/src/str/mod.rs:
```diff
@@ -213,7 +209,81 @@ impl str {
+        if index > self.len() {
+            slice_error_fail(self, index, index)
+        } else {
+            let upper_bound = (index + 4).clamp(0, self.len());
```
The lower bound on this clamp doesn't serve a purpose since `index + 4` is unsigned and always &gt;=0.

```suggestion
            let upper_bound = min(index + 4, self.len());
```

clarfonthey replied:
This was mostly for readability, since I assumed that this would always optimise the same and it was less prone to swapping a min for a max. Will change back if it seems necessary.

clarfonthey replied:
When I went through the code I ended up fixing this anyway, since in hindsight it's not that much more readable to use clamp.

dtolnay commented on library/core/src/str/mod.rs:
```diff
@@ -213,7 +209,81 @@ impl str {
+            slice_error_fail(self, index, index)
+        } else {
+            let upper_bound = (index + 4).clamp(0, self.len());
+            let new_index =
+                self.as_bytes()[index..upper_bound].iter().position(|b| b.is_utf8_char_boundary());
+
+            // this case should only be reached if `upper_bound == self.len()`
+            index + new_index.unwrap_or(upper_bound)
```
I'm not sure that this implementation is correct. For example consider `"x".ceil_char_boundary(1)`. According to the documentation this should be `1` because is_char_boundary(1) is true. But tracing through, we have:

- `index` is 1
- `self.len()` is 1
- `index > self.len()` is false so we take the `else` branch
- `index + 4` is 5
- `upper_bound` is 5.clamp(0, 1) which is 1
- `self.as_bytes()[1..1]` is empty slice
- `new_index` is None
- `new_index.unwrap_or(upper_bound)` is 1
- `index + new_index.unwrap_or(upper_bound)` is **2**(??)

clarfonthey replied:
You're right; `ceil_char_boundary` isn't done right. It should only add `index` to `Some(pos)` returned by `position`, and not `upper_bound`.

dtolnay reviewed (changes requested):
These method names and signatures look good to me. I was skeptical about the previous iterations but these are nice. Thanks for all the discussion above.

Please open a tracking issue and update the `unstable` attributes in the PR.

This is going to require more extensive test coverage prior to landing. See if you can find where the unit tests are for the other methods in this impl block.

crlf0710 commented:
@clarfonthey Ping from triage, would you mind addressing the comments above? Thanks!

clarfonthey commented:
Hey, sorry that I took so long getting to this. Will add some more tests and see what I can do about fixing the implementation.

clarfonthey commented:
Okay, this is finally ready for a proper review. I added a (hopefully) thorough set of tests and fixed the issue with `ceil_char_boundary`, which also did result in the code being more readable (IMHO) anyway.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #90414) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

scottmcm commented:
This is looking pretty good to me, code-wise, now.

Can you please add the tracking issue (as mentioned in https://github.com/rust-lang/rust/pull/86497#pullrequestreview-819834179)?  With that I think it'll be good to go.

clarfonthey commented:
Tracking issue made, also fixed the test being called `ceil_char_bonudary` instead of `ceil_char_boundary` 😳

Should be ready for final CR

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5097341052?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e7006705d0379e343d92a93cb31a3ed5284d1ee6/checks/5097341052/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({doc::src/doc/edition-guide}) not skipped for "bootstrap::doc::EditionGuide" -- not in [src/tools/tidy]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.17s
std/arch/index.html:153: broken link - `std/arch/x86/index.html`
std/arch/index.html:154: broken link - `std/arch/x86_64/index.html`
std/arch/index.html:155: broken link - `std/arch/arm/index.html`
std/arch/index.html:156: broken link - `std/arch/aarch64/index.html`
std/arch/index.html:157: broken link - `std/arch/riscv32/index.html`
std/arch/index.html:158: broken link - `std/arch/riscv64/index.html`
std/arch/index.html:159: broken link - `std/arch/mips/index.html`
std/arch/index.html:160: broken link - `std/arch/mips64/index.html`
std/arch/index.html:161: broken link - `std/arch/powerpc/index.html`
std/arch/index.html:162: broken link - `std/arch/powerpc64/index.html`
std/arch/index.html:163: broken link - `std/arch/nvptx/index.html`
std/arch/index.html:164: broken link - `std/arch/wasm32/index.html`
number of HTML files scanned: 32008
number of HTML redirects found: 9916
number of links checked: 2167980
number of links ignored due to external: 108226
```

</details>

clarfonthey commented:
(I had accidentally committed the wrong version of a submodule when rebasing, so, hopefully it doesn't fail this time.)

scottmcm commented:
Thanks for your persistence here!  These look good to go for nightly.

@bors r+

If anyone has further thoughts, please use the tracking issue #93743 or a new bug cc'ing it.

bors commented:
:pushpin: Commit edd318c313763d8c4cf3e8cc339f433832d6454a has been approved by `scottmcm`

<!-- @bors r=scottmcm edd318c313763d8c4cf3e8cc339f433832d6454a -->
<!-- homu: {"type":"Approved","sha":"edd318c313763d8c4cf3e8cc339f433832d6454a","approver":"scottmcm"} -->
